### PR TITLE
[SYCL][L0] Fix memory leak in USM prefetch

### PR DESCRIPTION
**Pull request by bso-intel**
L0 plugin increments the ref count of the dependency event twice in USM prefetch.
The fix is to remove the unnecessary retain call.

### CODE REVIEW

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -7294,11 +7294,6 @@ pi_result piextUSMEnqueuePrefetch(pi_queue Queue, const void *Ptr, size_t Size,
   // Lock automatically releases when this goes out of scope.
   std::lock_guard<std::mutex> lock(Queue->PiQueueMutex);
 
-  _pi_ze_event_list_t TmpWaitList;
-  if (auto Res = TmpWaitList.createAndRetainPiZeEventList(NumEventsInWaitList,
-                                                          EventWaitList, Queue))
-    return Res;
-
   // Get a new command list to be used on this call
```
Comment (bso-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:7297]: We already call createAndRetainPiZeEventList for the EventWaitList at line 7313 below.
This code unnecessarily increment the reference count twice, which leads to memory leak.

**Review by againull — APPROVED** [review]

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 1 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:7297]: In all other cases we create `TmpWaitList` before creating the real Event (and I don't remember why now). Please investigate/document and make your change in line with the rest of the code (presumably remove the other call to createAndRetainPiZeEventList, not this one)

**Review by smaslov-intel — CHANGES_REQUESTED** [review]

**Comment by bso-intel** [comment]
@smaslov-intel ,
Actually, it is better to keep line 7313 than this removed code.
Later when we need to cleanup the events in WaitList.collectEventsForReleaseAndDestroyPiZeEventList(EventsToBeReleased), we need the length of the ref-count-incremented events to call piEventRelease() properly.
This length of the ref-count-incremented events was set in TmpWaitList.createAndRetainPiZeEventList() on the TmpWaitList.
By keeping line 7313, we keep this length on the EventWaitList properly.

**Comment by smaslov-intel** [comment]
> 

See https://github.com/intel/llvm/pull/3081 for why `createAndRetainPiZeEventList` is needed before command-list is allocated.

**Comment by bso-intel** [comment]
Thanks @smaslov-intel.
I fixed as you requested.

**Comment by bso-intel** [comment]
Hold the review. It seems it is mixed up with some other changes.

**Comment by smaslov-intel** [comment]
> Hold the review. It seems it is mixed up with some other changes.

Yes, please update. And please also add comment to `createAndRetainPiZeEventList` that it must be run before a subsequent `getAvailableCommandList` so that there is no dead-lock from waiting unsubmitted events in an open batch.

**Comment by bso-intel** [comment]
Ok. Now it shows the correct changeset.
I added the comment, too.

**Review by smaslov-intel — APPROVED** [review]

**Pull request merged**